File: src/EBNF/ast/Node.php
<?php

/**
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @license http://www.gnu.org/licenses/ GNU General Public License
 * @author  Sven Strittmatter <ich@weltraumschaf.de>
 * @package ast
 */

namespace de\weltraumschaf\ebnf\ast;

use de\weltraumschaf\ebnf\visitor\Visitor;

/**
 * Interface of an AST node.
 *
 * @package ast
 */
interface Node {

    /**
     * Returns the name of a node.
     *
     * @return string
     */
    public function getNodeName();

    /**
     * Defines method to accept {@link Visitors}.
     *
     * Imlements {@link http://en.wikipedia.org/wiki/Visitor_pattern Visitor Pattern}.
     *
     * @param Visitor $visitor Object which visits te node.
     *
     * @return void
     */
    public function accept(Visitor $visitor);

    /**
     * Probes equivalence of itself against an other node and collects all
     * errors in the passed {@link Notification} object.
     *
     * @param Node         $other  Node to compare against.
     * @param Notification $result Object which collects all equivlanece violations.
     *
     * @return void
     */
    public function probeEquivalence(Node $other, Notification $result);
}

File: java/src/main/java/de/weltraumschaf/ebnf/ast/Node.java
package de.weltraumschaf.ebnf.ast;

import java.util.Map;

/**
 * Interface of an AST node.
 *
 * @author Sven Strittmatter <weltraumschaf@googlemail.com>
 */
public interface Node extends Visitable {

    /**
     * Returns the name of a node.
     *
     * @return
     */
    String getNodeName();

    /**
     * Probes equivalence of itself against an other node and collects all
     * errors in the passed {@link Notification} object.
     *
     * @param other  Node to compare against.
     * @param result Object which collects all equivalence violations.
     */
    void probeEquivalence(Node other, Notification result);

    /**
     * Returns the depth of the node.
     *
     * Nodes with no child have a depth of 1. Nodes with children have
     * the max depth of the children plus one. The depth is same as the
     * length of of the longest path in the tree.
     *
     * @return
     */
    int depth();

    NodeType getType();

    boolean hasParent();
    Node getParent();
    boolean hasAttributes();
    Map<String, String> getAttributes();
    boolean hasAttribute(String name);
    String getAttribute(String name);
    void setAttribute(String name, String value);
    boolean isType(NodeType checked);
}
